python/ray/data/_internal/execution/interfaces/physical_operator.py

```diff
@@ -200,16 +200,17 @@ def progress_str(self) -> str:
         """
         return ""
 
-    def num_outputs_total(self) -> Optional[int]:
-        """Returns the total number of output bundles of this operator, if known.
+    def num_outputs_total(self) -> int:
+        """Returns the total number of output bundles of this operator.
 
+        The value returned may be an estimate.
         This is useful for reporting progress.
         """
         if self._estimated_output_blocks is not None:
             return self._estimated_output_blocks
         if len(self.input_dependencies) == 1:
             return self.input_dependencies[0].num_outputs_total()
-        return None
+        raise NotImplementedError
```

Comment:
i think raising an `AttributeError` may be more appropriate in this case, since we are not missing the underlying implementation.